File: Habitica/src/com/habitrpg/android/habitica/ui/helpers/DataBindingUtils.java
package com.habitrpg.android.habitica.ui.helpers;

import android.databinding.BindingAdapter;
import android.databinding.DataBindingUtil;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.v7.widget.CardView;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.data5tream.emojilib.EmojiTextView;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.databinding.ValueBarBinding;
import com.squareup.picasso.Picasso;

/**
 * Created by Negue on 05.09.2015.
 */
public class DataBindingUtils {

    @BindingAdapter("bind:imageName")
    public static void loadImage(ImageView view, String imageName) {
        if (view.getVisibility() == View.VISIBLE) {
            Picasso.with(view.getContext())
                    .load("https://habitica-assets.s3.amazonaws.com/mobileApp/images/" + imageName + ".png")
                    .into(view);
        }
    }

    @BindingAdapter("bind:questImageName")
    public static void loadQuestImage(ImageView view, String imageName) {
        if (view.getVisibility() == View.VISIBLE) {
            Picasso.with(view.getContext())
                    .load("https://habitica-assets.s3.amazonaws.com/mobileApp/images/" + imageName + ".png")
                    .resizeDimen(R.dimen.quest_image_width, R.dimen.quest_image_height)
                    .into(view);
        }
    }

    @BindingAdapter("bind:cardColor")
    public static void setCardColor(CardView cardView, int color) {
        if (color > 0) {
            color = cardView.getResources().getColor(color);
        }
        cardView.setCardBackgroundColor(color);
    }

    @BindingAdapter("app:backgroundColor")
    public static void setBackgroundTintColor(CheckBox view, int color) {
        if (color > 0) {
            color = view.getResources().getColor(color);
        }
        ViewHelper.SetBackgroundTint(view, color);
    }

    @BindingAdapter("app:backgroundColor")
    public static void setBackgroundTintColor(Button view, int color) {
        if (color > 0) {
            color = view.getResources().getColor(color);
        }
        ViewHelper.SetBackgroundTint(view, color);
    }

    @BindingAdapter("app:backgroundColor")
    public static void setBackgroundTintColor(View view, int color) {
        if (color > 0) {
            color = view.getResources().getColor(color);
        }
        view.setBackgroundColor(color);
    }

    @BindingAdapter("app:foregroundColor")
    public static void setForegroundTintColor(TextView view, int color) {
        if (color > 0) {
            color = view.getResources().getColor(color);
        }
        view.setTextColor(color);
    }

    @BindingAdapter("android:layout_weight")
    public static void setLayoutWeight(View view, float weight) {
        ValueBarBinding value_bar = DataBindingUtil.findBinding(view);
        if (weight == 0.0f || weight == 1.0f || value_bar.getPartyMembers()) {
            LinearLayout.LayoutParams layout = (LinearLayout.LayoutParams) view.getLayoutParams();
            layout.weight = weight;
            view.setLayoutParams(layout);
        } else {
            LayoutWeightAnimation anim = new LayoutWeightAnimation(view, weight);
            anim.setDuration(1250);
            view.startAnimation(anim);
        }
    }

    @BindingAdapter("app:rounded_background")
    public static void setRoundedBackground(View view, int color) {
        Drawable drawable = view.getResources().getDrawable(R.drawable.layout_rounded_bg);
        drawable.setColorFilter(color, PorterDuff.Mode.MULTIPLY);
        if (Build.VERSION.SDK_INT < 16) {
            view.setBackgroundDrawable(drawable);
        } else {
            view.setBackground(drawable);
        }
    }

    @BindingAdapter("app:rounded_background_int")
    public static void setRoundedBackgroundInt(View view, int color) {
        if (color != 0) {
            setRoundedBackground(view, view.getResources().getColor(color));
        }
    }

    @BindingAdapter("parsemarkdown")
    public static void bindEmojiconTextView(EmojiTextView textView, CharSequence value) {
        if (value != null) {
            textView.setText(MarkdownParser.parseMarkdown(value.toString()));
        }
    }

    public static class LayoutWeightAnimation extends Animation {
        float targetWeight;
        float initializeWeight;
        View view;

        LinearLayout.LayoutParams layoutParams;

        public LayoutWeightAnimation(View view, float targetWeight) {
            this.view = view;
            this.targetWeight = targetWeight;

            layoutParams = (LinearLayout.LayoutParams) view.getLayoutParams();
            initializeWeight = layoutParams.weight;
        }

        @Override
        protected void applyTransformation(float interpolatedTime, Transformation t) {
            layoutParams.weight = initializeWeight + (targetWeight - initializeWeight) * interpolatedTime;

            view.requestLayout();
        }

        @Override
        public boolean willChangeBounds() {
            return true;
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/helpers/DataBindingUtils.kt
package com.habitrpg.android.habitica.ui.helpers

import android.content.Context
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.view.View
import android.view.animation.Animation
import android.view.animation.Transformation
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.core.content.res.ResourcesCompat
import coil.imageLoader
import coil.load
import coil.request.ImageRequest
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.extensions.setTintWith
import com.habitrpg.android.habitica.helpers.AppConfigManager
import java.util.*
import kotlin.collections.HashMap

fun ImageView.loadImage(imageName: String?, imageFormat: String? = null) {
    DataBindingUtils.loadImage(this, imageName, imageFormat)
}

object DataBindingUtils {

    fun loadImage(view: ImageView?, imageName: String) {
        loadImage(view, imageName, null)
    }

    fun loadImage(view: ImageView?, imageName: String?, imageFormat: String? = null) {
        if (view != null && imageName != null && view.visibility == View.VISIBLE) {
            val fullname = getFullFilename(imageName, imageFormat)
            if (view.tag == fullname) {
                return
            }
            view.tag = fullname
            view.load(BASE_IMAGE_URL + fullname)
        }
    }

    fun loadImage(context: Context, imageName: String, imageResult: (Drawable) -> Unit) {
        loadImage(context, imageName, null, imageResult)
    }

    fun loadImage(context: Context, imageName: String, imageFormat: String?, imageResult: (Drawable) -> Unit) {
        val request = ImageRequest.Builder(context)
            .data(BASE_IMAGE_URL + getFullFilename(imageName, imageFormat))
            .target {
                imageResult(it)
            }
            .build()
        context.imageLoader.enqueue(request)
    }

    fun getFullFilename(imageName: String, imageFormat: String? = null): String {
        val name = when {
            spriteSubstitutions.containsKey(imageName) -> spriteSubstitutions[imageName]
            FILENAME_MAP.containsKey(imageName) -> FILENAME_MAP[imageName]
            imageName.startsWith("handleless") -> "chair_$imageName"
            else -> imageName
        }
        return name + if (imageFormat == null && FILEFORMAT_MAP.containsKey(imageName)) {
            "." + FILEFORMAT_MAP[imageName]
        } else {
            ".${imageFormat ?: "png"}"
        }
    }

    fun setRoundedBackground(view: View, color: Int) {
        val drawable = ResourcesCompat.getDrawable(view.resources, R.drawable.layout_rounded_bg, null)
        drawable?.setTintWith(color, PorterDuff.Mode.MULTIPLY)
        view.background = drawable
    }

    class LayoutWeightAnimation(internal var view: View, internal var targetWeight: Float) : Animation() {
        private var initializeWeight: Float = 0.toFloat()

        private var layoutParams: LinearLayout.LayoutParams = view.layoutParams as LinearLayout.LayoutParams

        init {
            initializeWeight = layoutParams.weight
        }

        override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
            layoutParams.weight = initializeWeight + (targetWeight - initializeWeight) * interpolatedTime

            view.requestLayout()
        }

        override fun willChangeBounds(): Boolean = true
    }

    const val BASE_IMAGE_URL = "https://habitica-assets.s3.amazonaws.com/mobileApp/images/"
    private val FILEFORMAT_MAP: Map<String, String>
    private val FILENAME_MAP: Map<String, String>

    private var spriteSubstitutions: Map<String, String> = HashMap()
        get() {
            if (Date().time - (lastSubstitutionCheck?.time ?: 0) > 180000) {
                field = AppConfigManager(null).spriteSubstitutions()["generic"] ?: HashMap()
                lastSubstitutionCheck = Date()
            }
            return field
        }
    private var lastSubstitutionCheck: Date? = null

    init {
        val tempMap = HashMap<String, String>()
        tempMap["head_special_1"] = "gif"
        tempMap["broad_armor_special_1"] = "gif"
        tempMap["slim_armor_special_1"] = "gif"
        tempMap["head_special_0"] = "gif"
        tempMap["slim_armor_special_0"] = "gif"
        tempMap["broad_armor_special_0"] = "gif"
        tempMap["weapon_special_critical"] = "gif"
        tempMap["weapon_special_0"] = "gif"
        tempMap["shield_special_0"] = "gif"
        tempMap["Pet-Wolf-Cerberus"] = "gif"
        tempMap["armor_special_ks2019"] = "gif"
        tempMap["slim_armor_special_ks2019"] = "gif"
        tempMap["broad_armor_special_ks2019"] = "gif"
        tempMap["eyewear_special_ks2019"] = "gif"
        tempMap["head_special_ks2019"] = "gif"
        tempMap["shield_special_ks2019"] = "gif"
        tempMap["weapon_special_ks2019"] = "gif"
        tempMap["Pet-Gryphon-Gryphatrice"] = "gif"
        tempMap["Mount_Head_Gryphon-Gryphatrice"] = "gif"
        tempMap["Mount_Body_Gryphon-Gryphatrice"] = "gif"
        tempMap["background_clocktower"] = "gif"
        tempMap["background_airship"] = "gif"
        tempMap["background_steamworks"] = "gif"
        tempMap["Pet_HatchingPotion_Veggie"] = "gif"
        tempMap["Pet_HatchingPotion_Dessert"] = "gif"
        tempMap["Pet-HatchingPotion-Dessert"] = "gif"
        tempMap["quest_windup"] = "gif"
        tempMap["Pet-HatchingPotion_Windup"] = "gif"
        tempMap["Pet_HatchingPotion_Windup"] = "gif"
        tempMap["quest_solarSystem"] = "gif"
        FILEFORMAT_MAP = Collections.unmodifiableMap(tempMap)

        val tempNameMap = HashMap<String, String>()
        tempNameMap["head_special_1"] = "ContributorOnly-Equip-CrystalHelmet"
        tempNameMap["armor_special_1"] = "ContributorOnly-Equip-CrystalArmor"
        tempNameMap["head_special_0"] = "BackerOnly-Equip-ShadeHelmet"
        tempNameMap["armor_special_0"] = "BackerOnly-Equip-ShadeArmor"
        tempNameMap["shield_special_0"] = "BackerOnly-Shield-TormentedSkull"
        tempNameMap["weapon_special_0"] = "BackerOnly-Weapon-DarkSoulsBlade"
        tempNameMap["weapon_special_critical"] = "weapon_special_critical"
        tempNameMap["Pet-Wolf-Cerberus"] = "Pet-Wolf-Cerberus"
        FILENAME_MAP = Collections.unmodifiableMap(tempNameMap)
    }
}
